sympy/polys/numberfields/modules.py

```diff
@@ -0,0 +1,1242 @@
+            # The submodule is multiplied by an element of the containing module.
+            # We presume this means we want a new submodule of the containing module.
+            gens = [other * e for e in self.basis_element_pullbacks()]
+            return self.container.submodule_from_gens(gens, hnf=hnf, hnf_modulus=hnf_modulus)
+        elif self.is_compat_submodule(other):
+            # This case usually means you're multiplying ideals, and want another
+            # ideal, i.e. another submodule of the same container module.
+            alphas, betas = self.basis_element_pullbacks(), other.basis_element_pullbacks()
+            gens = [a * b for a in alphas for b in betas]
+            return self.container.submodule_from_gens(gens, hnf=hnf, hnf_modulus=hnf_modulus)
+        return NotImplemented
+
+    def __mul__(self, other):
+        return self.mul(other)
+
+    __rmul__ = __mul__
+
+    def _first_power(self):
+        return self.copy()
+
+
+def is_HNF(dm):
+    """
+    Say whether a DomainMatrix is in that special case of Hermite normal form,
+    in which the matrix is also square and of maximal rank. This is the only
+    case that concerns us here.
+    """
+    if dm.domain.is_ZZ and dm.is_square and dm.is_upper:
+        n = dm.shape[0]
+        for i in range(n):
+            d = dm[i, i].element
+            if d <= 0:
+                return False
+            for j in range(i + 1, n):
+                if not (0 <= dm[i, j].element < d):
+                    return False
+        return True
+    return False
+
+
+class HNF(Submodule):
+    """
+    A submodule such that:
+      * Its container is a PowerBasis.
+      * Its number of generators equals the degree of the minimal polynomial
+        of the PowerBasis.
+      * Its matrix is in square, maximal-rank, Hermite normal form.
+    """
+
+    def __init__(self, container, matrix, denom=1, mult_tab=None):
+        super().__init__(container, matrix, denom=denom, mult_tab=mult_tab)
+        assert isinstance(self.container, PowerBasis)
+        assert self.n == self.T.degree()
+        assert is_HNF(matrix)
+
+    @property
+    def T(self):
+        return self.container.T
+
+    def power_basis_elements(self):
+        return self.basis_element_pullbacks()
+
+    def pb_elt_from_poly(self, f):
+        """
+        Produce an element of the PowerBasis, representing a given polynomial
+        after reduction mod our defining minimal polynomial ``T``.
+
+        Parameters
+        ==========
+
+        f: Poly over :ref:`ZZ` in same var as our defining poly ``T``.
+
+        Returns
+        =======
+
+        PowerBasisElement
+
+        """
+        return self.container.element_from_poly(f)
+
+    def pb_elt_from_col(self, col, denom=1):
+        return self.container(col, denom=denom)
+
+
+class Order(HNF):
+    """
+    An order in a number field.
+
+    This is an HNF that includes unity.
```

Comment:
> then it's their fault if they get bad results?

If it's easy to provide the code that can check this then then the docstring here can point to the `is_order()` method and just explain that this will not be checked automatically.

Reply:
Okay, I looked at this one in many different ways. Ultimately, what the functions in `primes` call for is a `Submodule` representing not just an order, but the maximal order, and I don't know of any straightforward way to test for that, aside from essentially running the Round Two algorithm all over again, which would be crazy.

So I've implemented a middle-of-the-road approach:

1. these functions begin by checking some "formal preconditions" for being the right type of `Submodule`, which will at least raise excepetions on obviously ill-formed cases.

2. the boolean properties being checked are pre-set by the `round_two` function, so that, under standard usage, the checks are instantaneous or "free".

Meanwhile, the `Order`, `Ideal`, and `HNF` classes are gone. They weren't needed, so this is a good simplification.